### ISSUE TITLE
Including `sum()` in a `chebop` op definition.

### ISSUE THREAD

**Issue by nickhale** [chebfun/chebfun#1243]
I expected the following to work (it did in v4):

```
N = chebop(@(x,u) diff(u,2) + sum(u), [-1, 1], 1);
u = N\0;
```

but I get the following:

```
Error using  + 
Matrix dimensions must agree.
Error in operatorBlock>@(z)A.stack(z)+B.stack(z) (line 189)
            C.stack = @(z) A.stack(z) + B.stack(z);
Error in colloc/instantiate>instantiateOne (line 46)
    A = disc.source.stack(disc);
Error in colloc/instantiate (line 34)
    M = instantiateOne(disc, data);
Error in colloc/instantiate (line 30)
            M{j,k} = instantiate(discJK);
Error in chebDiscretization/matrix (line 44)
[A, S] = instantiate(disc);
Error in linop/linsolve (line 116)
        [A, P] = matrix(disc);
Error in chebop/solvebvpLinear (line 46)
del = linsolve(L, rhs, linpref);
Error in chebop/solvebvp (line 175)
    [u, info] = N.solvebvpLinear(L, rhs - residual, N.init, pref, displayInfo);
Error in  \  (line 13)
[varargout{1:nargout}] = solvebvp(varargin{:}); 
```

I suspect the functional resulting from the `sum(u)` term needs to be promoted to an operator so it can be added to the `diff(.,2)` operator. 

Edit: function --> functional

**Comment by asgeirbirkis** [comment]
I think this might be related to #1209, for which I have an upcoming pull request.

**Comment by asgeirbirkis** [comment]
@nickhale have you got an example where you want to use this functionality so that I can include it in a test?

**Comment by nickhale** [comment]
Not really. I guess the examples above would suffice (and probably non-linear versions, too)?

**Comment by Hadrien-Montanelli** [comment]
I've just discussed this issue with @aaustin141. I'd like to have this feature to solve the Ohta-Kawasaki with SPIN. To do so, I need to create a chebop (and discretize it) of the form:
```
L = chebop(@(u) diff(u,2) + u - sum(u))
```
As @asgeirbirkis and @nickhale mentioned above, Chebfun can't do this now because `sum` is treated as a `functionalBlock` (see `adchebfun/sum`) but should be an `operatorBlock` here. (Indeed, `sum` should be a `functionalBlock` when it appears in the boundary conditions.)

**Comment by nickhale** [comment]
We just pushed something to the [feature-promote](https://github.com/chebfun/chebfun/tree/feature-promote) branch which aims to implement this. It feels a little hacky though. In particular, I'm not quite sure of the correct place to (a) check for and (b) implement the promotion. (Currently I check in `adchebfun/plus` and implement in `functionalBlock/plus`.) Perhaps @asgeirbirkis or @tobydriscoll might have better ideas?

**Comment by Hadrien-Montanelli** [comment]
@nickhale, should we go ahead and issue a pull request?

**Comment by nickhale** [comment]
Sure.

**Comment by asgeirbirkis** [comment]
@nickhale , @Hadrien-Montanelli you might want to look into the unmerged branch [feature-functionalBlock-diag](https://github.com/chebfun/chebfun/tree/feature-functionalBlock-diag) which also does implements a promotion mechanism for `functionalBlocks`, although a slightly different one (converting a 1 x Inf operator to an Inf x Inf one, where the entries on the diagonal are those from the `functionalBlock`, all other entries being 0).

Since you simply owerwrite `F.stack` in your `functionalBlock/promote` method, I think you'll be fine not assigning the `diffOrder`, `isZero` and `isNotDiffOrInt` fields explicitly, so that implementation looks good to me at first sight at least.

Regarding where to check, we might go along the way of always allowing a 1 x Inf `functionalBlock` to be added to an Inf x Inf `operatorBlock` (which seems to be what Matlab is moving towards with vector and matrix operations), in which case the checking should be done in `functionalBlock/plus` rather than `adchebfun/plus` I believe. Otherwise, checking in `adchebfun/plus` is probably the correct place. I don't have a strong opinion about what approach to choose, but I agree that it'd be nice to reinstate this functionality.

**Comment by nickhale** [comment]
@asgeirbirkis - I moved the checking as you suggested. This seems much more natural.

I'm confused by your alternative promotion however. In what situation would `diag()` be the right thing to do?

**Comment by asgeirbirkis** [comment]
@nickhale The diag promotion was just an experiment (which I can't recall why I was making, so not expecting to issue a PR for it), so feel free ignore that code (just wanted to point out the relevant fields of functionalBlocks for the functionality you're introducing)!

----- Original Message -----
